projects/wallet-service-reliable-integration/0001-design.md

```diff
@@ -0,0 +1,205 @@
+https://xstate.js.org/docs/guides/actors.html
+
+
+#### WebSocketActor
+
+The `WebSocketActor` will initiate and maintain a connection with the full-node's WebSocket feed.
+
+We are using actors to have a bi-directional channel to the WebSocket feed from the machine, it will receive events and send those events to the machine, using the same type as received from the feed, so if new events are added in the future, they will be automatically emitted to the machine (and probably handled by the '\*' transition)
+
+
+## Services
+
+#### handleVertexAccepted
+
+This service handles the `NEW_VERTEX_ACCEPTED` event from the WebSocket feed, it means that the full-node's consensus mechanism accepted this transaction and added it to the blockchain, so we need to also include it in the wallet-service's database.
+
+These are the steps that need to be taken when processing a new transaction:
+
+1. Add the transaction to the `transaction` table
+2. Add the transaction outputs to the `tx_outputs` table
+3. Update the `tx_outputs` table, setting the `spent_by` column to the new transaction received if the `utxo` was used as an input.
+4. Calculate the transaction balance for each address involved.
+5. Update address balances (`address`, `address_balance`, `address_tx_history`) with the updated balances
+6. Update wallet tables (`wallet_balance`, `wallet_tx_history`) with the transaction and updated balances.
+7. Generate new addresses
+8. If this is a block, unlock all `utxos` that were heightlocked at this height.
+9. Use the transaction timestamp to unlock timelocked `utxos`
+10. If this is a token creation transaction, add the token information on the `token` table
+11. Increment the transactions count for each token involved in this transaction
+
+*We already have those methods well tested in the wallet-service lambdas, we can just migrate them to the daemon.*
+
+#### handleMetadataChanged
+
+This service handles the `VERTEX_METADATA_CHANGED` event from the WebSocket feed, it will be emitted from the full-node on different situations:
+
+1. The full-node consensus mechanism changed the metadata of this vertex before accepting it
+2. This transaction has been confirmed by a block
+3. This transaction has been voided
+
+We need to detect and handle this event differently for each situation. We can detect the situation type by inspecting our own database and checking what changed.
+
+`(1)`:
+If we don't have this transaction on the database, it's the situation (1). We can just add it to the database by invoking the `handleVertexAccepted` service.
+
+`(2)`:
+We can do an `UPSERT` in the `transaction` database updating the `height` column. No balance re-calculation is needed here.
```

Comment:
Why the `height`?

Reply:
Because that's what indicates to the wallet-service that the transaction has been confirmed by a block as we don't store metadata